### PR TITLE
ports/esp32/machine_timer.c: Solves #4078, ESP32 timer reinitilization error

### PR DESCRIPTION
**Pull request by dybber**
Solves the problem where multiple calls to `machine.Timer.init()` causes an `OSError 261` on ESP32 boards.

The fix is explained here: https://github.com/espressif/arduino-esp32/issues/1869#issuecomment-423016007

Solves issue #4078.

### CODE REVIEW

**Comment by dpgeorge** [comment]
Thanks for the patch, but I'm not sure it's the right way to solve that issue of error 261.  Active timers are put on a linked list (on the MicroPython side) and if you create a timer twice then it'll be placed twice on this list.  I think it would be better to check for an existing timer on the list when creating one, and reuse it if found.

**Comment by dybber** [comment]
Oh yes, didn't actually think of that! But now when checking the code, I don't believe that's the case. 

The init function calls machine_timer_disable which removes the existing timer from the list, before changing parameters and reenabling it.

https://github.com/micropython/micropython/blob/master/ports/esp32/machine_timer.c#L179

**Comment by dpgeorge** [comment]
> The init function calls machine_timer_disable which removes the existing timer from the list, before changing parameters and reenabling it.

But that doesn't stop there being multiple instances of Timer all pointing to the same underlying hardware resource, because these multiple Timer instances are distinct pointers to the MicroPython heap.

**Comment by dybber** [comment]
Okay, I see. I can implement a reference counting solution. Just as a sanity check before I start, do you think the following would be a suitable solution:

 - Timer's are added to the list already in the Timer constructor, not by the .init() method, if added twice reference counter is increased and the existing timer returned.
 - Timer's are stopped started by .init() / .deinit() method, but does not alter the list
 - Timer's are removed from the list in `__del__` (or -1 ref count)

Btw. I still believe the flags in the patch are necessary though, to solve the problem that occurs in programs such as this, which also raises OSError 261:
```python
timer0 = Timer(0)
timer0.init(period=1000, mode=Timer.PERIODIC, callback=my_callback)
timer0.deinit()
timer0.init(period=1000, mode=Timer.PERIODIC, callback=my_callback)
```

**Comment by dpgeorge** [comment]
> I can implement a reference counting solution.

I don't think it needs that.  There are only 4 timers in total so just keeping them in the link list forever after creation is fine (until soft reset, via `machine_timer_deinit_all()`).  Adding them to the list in the constructor sounds good.


>  I still believe the flags in the patch are necessary though, to solve the problem that occurs in programs such as this, which also raises OSError 261

That particular code runs fine for me, no error.

**Comment by dybber** [comment]
Okay, check the latest commit. 

You are right, the flags are not necessary. The example raising OSError 261 above only appeared when re-executing the code after a soft-reboot, however, it works now without the flags.

File: ports/esp32/machine_timer.c [hunk 1]
```diff
@@ -73,8 +73,13 @@ STATIC esp_err_t check_esp_err(esp_err_t code) {
...
-    while (MP_STATE_PORT(machine_timer_obj_head) != NULL) {
-        machine_timer_disable(MP_STATE_PORT(machine_timer_obj_head));
+    // Disable, deallocate and remove all timers from list
+    machine_timer_obj_t **t = &MP_STATE_PORT(machine_timer_obj_head);
+    while (*t != NULL) {
+        machine_timer_disable(*t);
+        machine_timer_obj_t **next = &(*t)->next;
+        m_del_obj(machine_timer_obj_t, *t);
+        t = next;
     }
```
Comment (dpgeorge) [inline ports/esp32/machine_timer.c]: Does this need a `MP_STATE_PORT(machine_timer_obj_head) = NULL` at the end?

Comment (dybber) [reply ports/esp32/machine_timer.c]: Good catch, my C is rusty after many years in the FP world. I have commited a small fix.

**Comment by dpgeorge** [comment]
Thank you!  Merged in de76f73d34a63fa34c520125f226e0c159ce0b10

**Pull request closed without merging**